### PR TITLE
Refactor Trick Room and fix interactions with speed-based moves

### PR DESCRIPTION
**Pull request by ascriptmaster**
Trick Room now uses a ModifyDecisionSpeed event to get its speed instead
of modifying the stat directly, since the Trick Room glitch would cause
calculations from Gyro Ball or Electro Ball to be inaccurate if the
Pokemon have speeds above and below the cutoff, respectively.

### CODE REVIEW

File: test/simulator/moves/trickroom.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+		battle = BattleEngine.Battle.construct('battle-trickroom-rollover', 'customgame');
+		battle.join('p1', 'Guest 1', 1, [{species: 'Ninjask', ability: 'swarm', moves: ['protect', 'spore']}]);
+		battle.join('p2', 'Guest 2', 1, [{species: 'Ninjask', ability: 'blaze', moves: ['spore', 'trickroom']}]);
+		battle.commitDecisions(); // Team Preview
+		battle.choose('p2', 'move 2');
+		battle.commitDecisions();
+
+		battle.on('ModifySpe', battle.getFormat(), function (spe, pokemon) {
+			// We are using abilities to decide who has what speed, exactly.
+			// In-game, this can be tested (albeit slightly less precicely) with Choice Scarfed Ninjasks with various speed boosts and EVs
```
Comment (Slayer95) [inline test/simulator/moves/trickroom.js]: I'd suggest transcripting the Ninjask test here.
Ideally, all our battle tests should be reproducible as-is in cartridge.

Comment (ascriptmaster) [inline test/simulator/moves/trickroom.js]: Transcribe the test, or put the actual test directly in here?

**Comment by ascriptmaster** [comment]
The test failed??? wtf. brb

File: test/simulator/moves/trickroom.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+		assert.strictEqual(battle.p1.active[0].speciesid, 'hippowdon');
+		assert.strictEqual(battle.p2.active[0].speciesid, 'ninetales');
+		assert.strictEqual(battle.effectiveWeather(), 'sunnyday');
+	});
+
+	// The following two tests involve the Trick Room glitch, where turn order changes when a Pokemon goes to 1809 speed.
+	// In these tests, we manually set the speed using events, but the effects can be approximately replicated with the following:
+	// 1808: Deoxys-Speed with 31 Speed IVs, 224 Speed EVs, neutral nature, with a +6 Speed boost.
+	// 1809: Ninjask with 31 Speed IVs, 184 Speed EVs, neutral nature, with a +4 Speed boost and a Choice Scarf.
+	// over 1810: Either of the above Pokemon with their Speed raised further (more EVs, boosts, Tailwind, etc.)
```
Comment (Slayer95) [inline test/simulator/moves/trickroom.js]: I forgot to reply, sorry.
I meant: feed this Ninjask to the simulator and check that what happens is the same as in cartridge.

Comment (ascriptmaster) [inline test/simulator/moves/trickroom.js]: They are fed into the simulator now.

**Comment by Zarel** [comment]
Ugh, don't add events if they're going to be used exactly once.

**Comment by ascriptmaster** [comment]
ugh I understand why but I couldn't think of anything better to do because I also hate patching in functions like we were doing

Fun fact, the old `getStatCallback` is also used exactly once, for Trick Room, but its location caused it to destroy Gyro Ball calcs and whatnot when the TR glitch was triggered, prompting this PR. (Also since turn order for mons above the cutoff)

**Comment by Zarel** [comment]
If you'd like, you can designate a section of `scripts.js` for this.

So in `scripts.js`:

`onModifyDecisionSpeed:` if TR `return (0x2710 - speed) & 0x1FFF`

and then 

			this.speed = this.onModifyDecisionSpeed(this.getStat('spe'));

Nice and fast, and not nearly as patchy as otherwise.

Maybe we can add `this.scriptedEvent` to handle it...

**Comment by ascriptmaster** [comment]
This should work properly.

There are some other rules for speed based on what I remember from @Marty-D's test results, but I'll need him to verify that what I'm doing is actually correct.

**Comment by Marty-D** [comment]
I haven't had time to look this over thoroughly, but you seem to have covered everything from the tests. :+1:

**Comment by ascriptmaster** [comment]
Marty, please look at the decision speed function and make sure that's correct at least

**Comment by Marty-D** [comment]
That's actually the only part I've looked at; seems fine.

**Comment by ascriptmaster** [comment]
Good, that was the only part I absolutely needed you to verify.

**Pull request merged**